src/libraries/Microsoft.Win32.SystemEvents/src/Microsoft/Win32/SystemEvents.cs

```diff
@@ -1075,43 +1070,59 @@ private static void RemoveEventHandler(object key, Delegate? value)
 
         private static void Shutdown()
         {
-            if (s_systemEvents != null)
+            if (s_systemEvents is null)
             {
-                lock (s_procLockObject)
+                return;
+            }
+
+            lock (s_procLockObject)
+            {
+                if (s_systemEvents is null)
+                {
+                    return;
+                }
+
+                // If we are using system events from another thread, request that it terminate
+                if (s_windowThread is not null)
                 {
-                    if (s_systemEvents != null)
-                    {
-                        // If we are using system events from another thread, request that it terminate
-                        if (s_windowThread != null)
-                        {
 #if DEBUG
-                            unsafe
-                            {
-                                int pid;
-                                int thread = Interop.User32.GetWindowThreadProcessId(s_systemEvents._windowHandle, &pid);
-                                Debug.Assert(thread != Interop.Kernel32.GetCurrentThreadId(), "Don't call Shutdown on the system events thread");
+                    unsafe
+                    {
+                        int pid;
+                        int thread = Interop.User32.GetWindowThreadProcessId(s_systemEvents._windowHandle, &pid);
+                        Debug.Assert(thread != Interop.Kernel32.GetCurrentThreadId(), "Don't call Shutdown on the system events thread");
 
-                            }
+                    }
 #endif
-                            // The handle could be valid, Zero or invalid depending on the state of the thread
-                            // that is processing the messages. We optimistically expect it to be valid to
-                            // notify the thread to shutdown. The Zero or invalid values should be present
-                            // only when the thread is already shutting down due to external factors.
-                            if (s_systemEvents._windowHandle != IntPtr.Zero)
-                            {
-                                Interop.User32.PostMessageW(s_systemEvents._windowHandle, Interop.User32.WM_QUIT, IntPtr.Zero, IntPtr.Zero);
-                                GC.KeepAlive(s_systemEvents);
-                            }
+                    // The handle could be valid, Zero or invalid depending on the state of the thread
+                    // that is processing the messages. We optimistically expect it to be valid to
+                    // notify the thread to shutdown. The Zero or invalid values should be present
+                    // only when the thread is already shutting down due to external factors.
+                    if (s_systemEvents._windowHandle != IntPtr.Zero)
+                    {
+                        Interop.User32.PostQuitMessage(0);
```

Comment:
Made this change because docs mention to call `PostQuitMessage` instead of PostMessage WM_QUIT https://learn.microsoft.com/en-us/windows/win32/winmsg/wm-quit#remarks

Reply:
@AustinWise is correct. The principal difference between `PostMessage(..., WM_QUIT, ...)` and `PostQuitMessage` is that the first one actually queues a message while the other one sets a flag on the message queue. From user's perspective the difference is the order in which `GetMessage`/`PeekMessage` returns the message and few quirks on how it behaves for nested message loops. Neither of that is a concern for the SystemEvents thread that is supposed to be entirely under our control and running a single non-nested message loop.

Reply:
Appreciate the extra eyes here, that is good catch! Will update to make sure this is called on the SystemEvents thread instead.